skimage/morphology/_restorable_queue.pxi

```diff
@@ -0,0 +1,119 @@
+"""FIFO queue whose items can be iterated multiple times.
+
+.. warning:: Expects "QueueItem" to be defined before inclusion!
+
+This queue can be operated like a class. The structure stores the state of the
+instance and the contained functions act as instance methods. The important
+distinction compared to normal queues is that popping an element doesn't remove
+it internally. Thus unless the queue's internal buffer is explicitly cleared
+(see `queue_clear`) popped items can be restored at anytime using
+`queue_restore`.
```

Comment:
Ah, ok. I might rename this a "QueueWithHistory"? The main purpose is to ensure that we have a "list" at the end of everything that was ever in the queue, so that as we proceed with floodfill, if we need to undo the whole thing, we can? To clarify, this is not to "backtrack" individual elements, right?

In this case, then (a) "QueueWithHistory" makes more sense to me, and (b) please add these notes explaining what you want to do with it. Yes, it's a good solution. =)

Reply:
Sounds good. I don't have a strong opinion about the current name. 

> The main purpose is to ensure that we have a "list" at the end of everything that was ever in the queue, so that as we proceed with floodfill, if we need to undo the whole thing, we can?

Correct!

> To clarify, this is not to "backtrack" individual elements, right?

No, although the insertion order is maintained and backtracking (if you mean iterating over the plateau values in reverse insertion order) would be possible with a few tweaks, I don't need to maintain order. Its just a consequence of the current implementation. 

I'll add the explanation in the header of the file.